### ISSUE TITLE
BUG?: rendering of parts in weird order

### ISSUE THREAD

**Issue by MatteoPiovanelli-Laser** [OrchardCMS/Orchard#7376]
From a conversation on Gitter between me and @sebastienros 

MatteoPiovanelli-Laser @MatteoPiovanelli-Laser ott 25 12:23
Can I control the order in which the parts are rendered? I meant, the order in time, rather than position.
I have this part that replaces the Layout.Title. This used to work correctly. Now in 1.10 it looks like TitleParts are being rendered after it, rather than before as they used to in 1.8, so they reset the Layout.Title back to its "standard" value. There are reasons why we want to do that, one of them being e.g. the fact that the TitlePart.Title is used to generate permalinks.

Sébastien Ros @sebastienros ott 25 19:05
@MatteoPiovanelli-Laser so many questions, not sure were to start
[...]
For the title part issue, maybe we fixed a bug that resulted in this changed behavior, safe thing to do is file an issue with a repro. But the idea is that if you feature "depends on" the title feature, then its handlers will be called after the title's ones.

So, I tried ensuring that that dependency was there.

The issue however appears to be more complex than that.

My part, which we'll call MyTitlePart, is used to replace the page's title. That is the only thing it does. For example, in a Page contentItem, you would have the content's title, and this additional field to insert the title that will appear on top of the browser's page.
`@Layout.Title = Model.TitleOverride;`
Easy.
I follow this steps.
 - Create a tenant. It will have the default "Welcome to Orchard!" Page.
 - Enable the feature containing MyTitlePart, and add MyTitlePart to the Page ContentType.
 - Edit that initial page and input any TitleOverride.
 - Viewing that page, the page's title is still "Welcome to Orchard!".
 - Create a new Page ContentItem. Input a "Real title" as the title for the TitlePart, and "Override" as the TitleOverride for MyTitlePart.
 - View the page you just created. The layout's title changed, as we wished, to "Override".

To summarize:
I only get the expected behaviour for content items I created AFTER enabling the feature and adding the part in the ContentType definition. For other items, even though the MyPartTitle driver and view are being executed (the debugger goes through) I end up with the default page's title.
It's as if the rendering order is different in those two cases, which is really odd.

Can someone pitch in and shed light on this? It might be I did something wrong, but I tried a couple times with the same result.

Best,
Matteo

**Comment by sebastienros** [comment]
Talked with @MatteoPiovanelli-Laser offline, this is the suggested solution

```
public class MyTitlePartHandler : ContentHandler {

        public MyTitlePartHandler() {
            OnIndexing<ITitleAspect>((context, part) => context.DocumentIndex.Add("title", part.Title).RemoveTags().Analyze());
            OnGetContentItemMetadata<MyTitlePart>((c, part) => c.Metadata.DisplayText = part.MyTitle);
        }
    }
```

**Comment by MatteoPiovanelli-Laser** [comment]
@sebastienros 
I tried adding the `OnGetContentItemMetadata` implementation suggested. That does not seem to do what I wish (suppose that for an example of ContentItem I have TitlePart.Title = "Title" and MyTitlePart.TitleOverride = "Override"):
 - If I remove the line `@Layout.Title = Model.TitleOverride;` from the view of the MyTitlePart, the Layout.Title takes its value from the ContentItem's TitlePart, i.e. the tag in the resulting html is `<title>Title</title>`
 - The handler like that has unwanted side effects related to the other places where DisplayText comes into play. For example, the first thing I notice was that it changed the "title" for the ContentItem in the list of contents (DisplayType == SummaryAdmin): I would normally see the link to the editor view with "Title", but by changing the metadata in the handler that would become "Override". That is not desired for this module/part (but it's good to know for future reference).

I did some more tests and experiments to figure out why I saw the behaviour described in my post above, and I found the reason for it: the LayoutPart.
In my "old" contentItems I had been fiddling with the LayoutPart (remember, I am using the Page ContentType for these tests). Namely, I had added the ContentItem's TitlePart into the Layout. The result of this seems to be that the copy of the TitlePart that is in the Layout gets rendered after the other parts in the item, with the result (in my specific case) that the Layout.Title gets overwritten.
As further evidence, I tried the following (without the suggested change in the handler):
 - Enabled my feature, and added MyTitlePart to the content definition of Page.
 - Create a first page (call it Page1), setting TitlePart.Title = "Title1" and MyTitlePart.TitleOverride = "Override1".
 - View the "Detail" view of Page1: its TitlePart displays correctly; in the html header I have `<title>Override1</title>`, that is the desired outcome.
 - From Page1's editor view, add anywhere in its layout the TitlePart, and publish.
 - View the "Detail" view of Page1:  its TitlePart displays correctly; in the html header I have `<title>Title1</title>`, that is not the desired outcome.
 - Create another page (call it Page2), setting TitlePart.Title = "Title2" and MyTitlePart.TitleOverride = "Override2".
 - In the layout part of Page2, add anywhere the Content Item Page1.
 - View the "Detail" view of Page2: it will display things as expected; in the html header you will have `<title>Title1</title>`, that is the same we last had on Page1.
 - Go back to editing Page1: remove the TitlePart from its LayoutPart. Page1's "Detail" view will be back to its desired outcome, with the html header `<title>Override1</title>`.
 - View Page2's "Detail" view: in the html header there will be `<title>Override1</title>`.

It looks like the contents in the Layout part are rendered after everything else, so in my case they end up overriding the behaviour I am trying to implement.

I am not sure this is desired behaviour for LayoutParts. I mean, I think I understand why it's happening, but I can't think of a scenario where we would want something we add in the LayoutPart to override what we do with the Parts that we have in the item.

Best,
Matteo

**Comment by sebastienros** [comment]
@sfmskywalker 
The issue here is that apparently, when a TitlePart element is added to a Layout, the TitlePart handler is run again, is that expected?
